### PR TITLE
MEAN specs

### PR DESCRIPTION
**Pull request by tejasmanohar**

### CODE REVIEW

File: specs/haproxy/Dockerfile [hunk 1]
```diff
@@ -1,11 +1,8 @@
 From haproxy:1.6.4
 
-COPY haproxy.cfg /usr/local/etc/haproxy/haproxy.cfg
```
Comment (tejasmanohar) [inline specs/haproxy/Dockerfile]: Didn't see a reason to keep this around

File: specs/mean/service/config/database.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+    remoteUrl : 'mongodb://node:nodeuser@mongo.onmodulus.net:27017/uwO3mypu',
+    localUrl: process.env.MONGO_URI || 'mongodb://localhost/meanstacktutorials'
```
Comment (tejasmanohar) [inline specs/mean/service/config/database.js]: only change i had to make to scotch repo, maybe i should just PR them?

**Comment by tejasmanohar** [comment]
:\ cov check failed. Should we be
- writing tests for specs
- ignoring this
- adding unrelated tests
?

**Comment by tejasmanohar** [comment]
(@mmzhao is reviewing)

**Comment by tejasmanohar** [comment]
(@ejj - I moved example app to [tejasmanohar/node-todo](https://github.com/tejasmanohar/node-todo) & added some tests to improve cov (IMO we should exclude specs/ from cov and maybe even move to another repo but that fails checks _now_ so maybe later)

**Comment by tejasmanohar** [comment]
(edit: @LuiseV is reviewing)

**Comment by kklin** [comment]
I want to do a pass over this too -- after @LuiseV, before @ejj

File: specs/mongo/mongo.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
+var image = "tejasmanohar/quilt-mongo";
+var PORT = 27017;
+
+function Mongo(nWorker) {
+  // The "initial" container is the node the replica set is initialized from
+  // Initializing a replica set triggers an election of primary (master),
+  // secondaries (slaves), etc.
+  var initialContainer = new Container(image);
+  initialContainer.env["ROLE"] = "initial";
```
Comment (LuiseV) [inline specs/mongo/mongo.js]: Should use `setEnv()` - which you actually do already in line 23. Is there a reason to set this variable twice?

Comment (tejasmanohar) [inline specs/mongo/mongo.js]: oh i'll fix that, i originally used .env because setEnv accepted a full object & it was easier to read than accumulating results and calling setEnv at end

File: specs/mongo/mongo.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  this._initial.connect(PORT, this._subsequents);
+  this._subsequents.connect(PORT, this._initial);
+  this._subsequents.connect(PORT, this._subsequents);
+};
+
+Mongo.prototype.deploy = function(deployment) {
+  deployment.deploy([this._initial, this._subsequents]);
+};
+
+Mongo.prototype.deploy = function(deployment) {
```
Comment (LuiseV) [inline specs/mongo/mongo.js]: What's the purpose of this function? When do we use `this.initial` over `this._initial`?

Comment (tejasmanohar) [inline specs/mongo/mongo.js]: eek, looks like a typo when merging stuff. fixed. i'll run through the diff another time

File: specs/mongo/mongo.js (hunk 3 above)
Comment (tejasmanohar) [inline specs/mongo/mongo.js]: fixed

File: specs/mongo/mongo.js (hunk 4 above)
Comment (tejasmanohar) [inline specs/mongo/mongo.js]: we should get a linter! // cc @kklin

File: specs/mongo/mongo.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
+var image = "tejasmanohar/quilt-mongo";
+var PORT = 27017;
+
+function Mongo(nWorker) {
+  // The "initial" container is the node the replica set is initialized from
+  // Initializing a replica set triggers an election of primary (master),
+  // secondaries (slaves), etc.
+  var initialContainer = new Container(image);
```
Comment (LuiseV) [inline specs/mongo/mongo.js]: It seems like the MongoDB terminology for replication is `primary` and `secondary`, so we might want to use that over `initial` and `subsequent`.

File: specs/mongo/mongo.js (hunk 4 above)
Comment (kklin) [inline specs/mongo/mongo.js]: Don't know much about how people generally use linters. Would the linter be built into the language, or just a tool that you invoke with `make` or something?

File: specs/mongo/mongo.js (hunk 5 above)
Comment (tejasmanohar) [inline specs/mongo/mongo.js]: I avoided `primary` and `secondary`, as the node you instantiate the replica set on isn't necessarily the `primary`-- creating a replica set triggers an election (can be re-triggered automatically by Mongo or manually). All nodes can potentially serve all roles.

Comment (tejasmanohar) [inline specs/mongo/mongo.js]: I believe the node you instantiate the replica set on is preferred as the primary though so I'm ok changing the wording if you think it's confusing... or simply adding a comment, I just wanted to make sure no one makes an assumption that the "initial" node *is* the primary

File: specs/mongo/mongo.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+Mongo.prototype.connect = function(port, to) {
+  to.services().forEach(function(service) {
+    this._initial.connect(port, service);
+    this._subsequents.connect(port, service);
+  }.bind(this));
+};
+
+Mongo.prototype.uri = function(dbName) {
+  var services = [this._initial].concat([this._subsequents]);
+  var hostnames = _.flatten(services.map(function(label) {
```
Comment (LuiseV) [inline specs/mongo/mongo.js]: Might want to use `service` instead of `label` now that we replaced `Label`

File: specs/mongo/mongo.js (hunk 5 above)
Comment (LuiseV) [inline specs/mongo/mongo.js]: Oh, that makes sense. You should stick with then current  naming then! Maybe you can add a comment saying that the initial is not necessarily the primary, so we avoid confusion like mine, but that's up to you!

File: specs/haproxy/haproxy.js [hunk 7]
```diff
@@ -1,15 +1,29 @@
...
+    var addresses = hostnames.map(function(host) {
+      return host + ":" + port;
+    });
+    var hapRef = new Container(image).replicate(n);
+    hapRef.forEach(function(c) {
+      c.setEnv("ADDRS", addresses.join(","));
+    });
+
+    this.service = new Service("hap", hapRef);
+    services.forEach(function(label) {
```
Comment (tejasmanohar) [inline specs/haproxy/haproxy.js]: fixed

**Comment by LuiseV** [comment]
Should we generally push the images to the Quilt repo on docker hub instead of your repo, @tejasmanohar ?

**Comment by ejj** [comment]
Yep, we should push to docker hub.  Ideally we could setup an auto-build

**Comment by LuiseV** [comment]
By the way, this was not a thorough review - just the things I stumbled upon when using the specs. As we agreed on, @kklin should do the proper review before @ejj takes over.

File: specs/haproxy/Dockerfile [hunk 8]
```diff
@@ -1,11 +1,8 @@
...
-# This is just creating a user 'haproxy' with password 'haproxy'
-RUN useradd haproxy -p '$6$xdWkk9b.$lQiYLBsaVIxujlXj2lAhbNskglkrUcHKjajfiaIMwHOVxGrGhC30oGhBi939PFOM9DmbfgdTCqUjrQUc2GIAC0' && \
-mkdir -m 000 /haproxy
+RUN apt-get update && apt-get install -y curl sockstat lsof vim
```
Comment (kklin) [inline specs/haproxy/Dockerfile]: Why do we need these dependencies?

File: specs/mean/example.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
+var HaProxy = require("github.com/NetSys/quilt/specs/haproxy/haproxy").Haproxy;
```
Comment (kklin) [inline specs/mean/example.js]: Add to `spec_test.go`

File: specs/mean/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,10 @@
+FROM node:6.8
+
+COPY scripts/run /
```
Comment (LuiseV) [inline specs/mean/Dockerfile]: What is this `run` file? Should it be `run.go`?

File: specs/mongo/example.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+var namespace = createDeployment({
+    namespace: "CHANGE_ME",
+    adminACL: ["local"],
+});
+var baseMachine = new Machine({
+    provider: "Amazon",
+    cpu: new Range(2),
+    ram: new Range(2),
+    sshKeys: githubKeys("CHANGE_ME"),
```
Comment (tejasmanohar) [inline specs/mongo/example.js]: that's a bug, right?

File: specs/mean/Dockerfile (hunk 10 above)
Comment (tejasmanohar) [inline specs/mean/Dockerfile]: it's the binary

File: specs/haproxy/Dockerfile (hunk 8 above)
Comment (tejasmanohar) [inline specs/haproxy/Dockerfile]: whoops, i used them to debug :P

Comment (tejasmanohar) [inline specs/haproxy/Dockerfile]: fixed

File: specs/mean/example.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    sshKeys: githubKeys("CHANGE_ME"),
+});
+namespace.deploy(baseMachine.asMaster());
+namespace.deploy(baseMachine.asWorker().replicate(3));
+
+var mongo = new Mongo(3);
+var app = new App(3, 8080, { MONGO_URI: mongo.uri("mean-example") });
+var haproxy = new HaProxy(2, app.services(), 8080);
+
+mongo.connect(27017, app);
```
Comment (tejasmanohar) [inline specs/mean/example.js]: done

File: specs/mongo/scripts/run.go [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+			log.Fatalf("Error running Mongo server: %s", err.Error())
+		}
+	}()
+
+	if role == initialRole && len(peers) > 0 {
+		checkConnection(fmt.Sprintf("%s:27017", host))
+
+		log.Printf("Setting up replica set with peers: %+v", peers)
+		wg.Add(1)
+		go func() {
```
Comment (tejasmanohar) [inline specs/mongo/scripts/run.go]: command#Run blocks

File: specs/mean/Dockerfile (hunk 10 above)
Comment (LuiseV) [inline specs/mean/Dockerfile]: Oops, yeah, of course!

File: specs/mongo/mongo.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+  this._subsequents.connect(PORT, this._initial);
+  this._subsequents.connect(PORT, this._subsequents);
+};
+
+Mongo.prototype.deploy = function(deployment) {
+  deployment.deploy([this._initial, this._subsequents]);
+};
+
+Mongo.prototype.services = function() {
+  return [this._initial].concat(this._subsequents);
```
Comment (tejasmanohar) [inline specs/mongo/mongo.js]: oops yeah i had more stuff before

File: specs/mean/app.js [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  this._app.connect(this._port, this._app);
+}
+
+App.prototype.deploy = function(deployment) {
+  deployment.deploy([this._app]);
+};
+
+App.prototype.services = function(customPort) {
```
Comment (LuiseV) [inline specs/mean/app.js]: Redundant arg

File: specs/mongo/mongo.js [hunk 16]
```diff
@@ -0,0 +1,60 @@
+var image = "tejasmanohar/quilt-mongo";
```
Comment (tejasmanohar) [inline specs/mongo/mongo.js]: yeah @ejj

File: specs/mongo/example.js [hunk 17]
```diff
@@ -0,0 +1,19 @@
+var Mongo = require("github.com/NetSys/quilt/specs/mongo/mongo");
```
Comment (tejasmanohar) [inline specs/mongo/example.js]: added mean itself

File: specs/mongo/example.js (hunk 11 above)
Comment (tejasmanohar) [inline specs/mongo/example.js]: fair, fixed

File: specs/mongo/example.js [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+var baseMachine = new Machine({
+    provider: "Amazon",
+    cpu: new Range(2),
+    ram: new Range(2),
+    sshKeys: githubKeys("CHANGE_ME"),
+});
+var mongo = new Mongo(nWorker);
+
+namespace
+  .deploy(baseMachine.asMaster())
```
Comment (tejasmanohar) [inline specs/mongo/example.js]: ooh, will fix!

File: specs/mongo/mongo.js [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+Mongo.prototype.connect = function(port, to) {
+  to.services().forEach(function(service) {
+    this._initial.connect(port, service);
+    this._subsequents.connect(port, service);
+  }.bind(this));
+};
+
+Mongo.prototype.uri = function(dbName) {
+  var services = [this._initial].concat([this._subsequents]);
+  var hostnames = _.flatten(services.map(function(service) {
```
Comment (tejasmanohar) [inline specs/mongo/mongo.js]: it's valid, i'm iffy that it's nicer but i changed it for now :)

File: specs/mongo/scripts/run_test.go [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestBuildHAProxyConfig(t *testing.T) {
+	ip := []string{"0.0.0.0", "127.0.0.1", "192.168.52.7"}
+	assert.EqualValues(t, buildMongoConfig(ip), mongoConfig{
```
Comment (tejasmanohar) [inline specs/mongo/scripts/run_test.go]: tbh i added this test so cov would pass. i don't see much value in testing that stdlib yaml marshal works but if you'd like, sure

File: specs/mongo/mongo.js (hunk 19 above)
Comment (tejasmanohar) [inline specs/mongo/mongo.js]: map passes an argument

Comment (tejasmanohar) [inline specs/mongo/mongo.js]: well, multiple arguments in fact so i avoid passing externally defined functions to map

File: specs/haproxy/.gitignore [hunk 21]
```diff
@@ -0,0 +1 @@
+scripts/start
```
Comment (tejasmanohar) [inline specs/haproxy/.gitignore]: lol :P done

File: specs/mongo/scripts/run.go [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+func executeMongoCommand(op string) error {
+	cmd := exec.Command("mongo", "--eval", op)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+func resolveHostname(hostname string) []string {
```
Comment (tejasmanohar) [inline specs/mongo/scripts/run.go]: everywhere else it's used-- it's a hack until docker network plugin is up. i think we should abstract it if it's still necessary once that's in

File: specs/mongo/mongo.js (hunk 5 above)
Comment (tejasmanohar) [inline specs/mongo/mongo.js]: fair

File: specs/mongo/mongo.js (hunk 4 above)
Comment (tejasmanohar) [inline specs/mongo/mongo.js]: just a tool like eslint. shouldn't be a part of quilt but should be used for quilt specs

**Comment by tejasmanohar** [comment]
ooh build-push doesn't like sub-vendor folder

File: specs/mean/app.js [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+function App(nWorker, port, env) {
+  this._port = port || 80;
+
+  var containers = new Container(image).replicate(nWorker);
+  this._app = new Service("app", containers);
+
+  var hosts = this._app.children();
+  for (var i = 0; i < containers.length; i++) {
+    containers[i].setEnv("HOST", hosts[i]);
+    _.each(env, function(v, k) {
```
Comment (tejasmanohar) [inline specs/mean/app.js]: :thumbs_up:

**Comment by tejasmanohar** [comment]
@kklin @LuiseV addressed nits, fixed build :)

File: specs/mongo/mongo.js [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+  // The "initial" container is the node the replica set is initialized from.
+  // Initializing a replica set triggers an election of primary (master),
+  // secondaries (slaves), etc. The initial container is not necessarily the primary.
+  var initialContainer = new Container(image);
+  initialContainer.setEnv("ROLE", "initial");
+
+  this._initial = new Service("mongo-initial", [initialContainer]);
+  initialContainer.setEnv("HOST", this._initial.hostname());
+
+  // "subsequent" containers are peers of the "initial" container. They receive the
```
Comment (ejj) [inline specs/mongo/mongo.js]: Btw, what happens if the initial container dies, and is restarted somewhere else?  Will mongo freak out?  Not sure we have the tools to deal with this well yet, but I'm curious about the scope of the problem.  We should discuss it and at least make a note of it.

File: specs/mongo/mongo.js [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+  // starting replica set configuration from the initial container.
+  var subsequentContainers = new Container(image).replicate(nWorker-1);
+  this._subsequents = new Service("mongo-subsequent", subsequentContainers);
+
+  var subsequentHosts = this._subsequents.children();
+  for (var i = 0; i < subsequentContainers.length; i++) {
+    subsequentContainers[i].setEnv("ROLE", "subsequent");
```
Comment (ejj) [inline specs/mongo/mongo.js]: Btw, what's accepted javascript style now?  Is there some sort of linter we can run?

File: specs/mongo/mongo.js [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+
+  // All containers should be able to communicate in/out, as they could all
+  // serve either role (primary, secondary), given a re-election upon failure, etc.
+  this._initial.connect(this._port, this._initial);
+  this._initial.connect(this._port, this._subsequents);
+  this._subsequents.connect(this._port, this._initial);
+  this._subsequents.connect(this._port, this._subsequents);
+};
+
+Mongo.prototype.deploy = function(deployment) {
```
Comment (ejj) [inline specs/mongo/mongo.js]: This is all *very* clean btw.  I like it a lot.

File: specs/mongo/example.js (hunk 17 above)
Comment (kklin) [inline specs/mongo/example.js]: Testing mean won't exercise this spec, though.

File: specs/mongo/mongo.js (hunk 25 above)
Comment (tejasmanohar) [inline specs/mongo/mongo.js]: i used a for loop here because i wanted the index to lookup in `subsequentHosts`, i think using the iteration prototype functions is "idiomatic". unfortunately, most of the linters are super configurable in the JS community & they don't really get into overall code style, mostly nits

Comment (tejasmanohar) [inline specs/mongo/mongo.js]: i agree we should adopt a linter for our own specs just to avoid stupid mistakes though we should prob leave that for another PR

File: specs/mongo/mongo.js (hunk 24 above)
Comment (tejasmanohar) [inline specs/mongo/mongo.js]: with mongo, this _should_ not be a problem (in the general / perfect case at least-- i'm sure there are more cases that could bite us so i'll keep thinking), and the failure would simply trigger a new election for master (in a new replica set if the initial dies). all apps connect to the replica set, not a specific node.

Comment (ejj) [inline specs/mongo/mongo.js]: I guess the larger point is, in your opinion, if a node comes up that things it's the "initial" even though a cluster is already running, mongo should deal with that gracefully.  If that sounds right I'm happy.

File: specs/mongo/mongo.js [hunk 27]
```diff
@@ -0,0 +1,66 @@
+var image = "tejasmanohar/quilt-mongo";
```
Comment (ejj) [inline specs/mongo/mongo.js]: The reason it makes sense, is someone could make their own container that's based on ours as the starting point.

File: specs/mongo/mongo.js (hunk 24 above)
Comment (tejasmanohar) [inline specs/mongo/mongo.js]: yep, but i do think it would make sense to be able to separate what should run if something is _rebooted_ vs. booted the first time, as some of the steps in run.go are unnecessary the second time around (e.g. creating a new replica set)

**Comment by tejasmanohar** [comment]
Pushed everything to quilt/* on DockerHub.

File: specs/haproxy/haproxy.js [hunk 28]
```diff
@@ -1,15 +1,28 @@
-var image = "quilt/haproxy"
+var image = "quilt/haproxy";
 var cfg = "/usr/local/etc/haproxy/haproxy.cfg";
 
-function Haproxy(n, hosts) {
-    var hostnames = hosts.children().join(",");
-    var hapRef = new Container(image, [hostnames, "haproxy", "-f", cfg]);
+function Haproxy(n, services, port) {
```
Comment (ejj) [inline specs/haproxy/haproxy.js]: You know, I wonder if we should add a function comment for this (and the other example specs in this patch).  Could help users avoid reading the code to figure out what the arguments mean.

File: specs/haproxy/scripts/start.go [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+		log.Fatalf("Error configuring HAProxy: %s", err.Error())
+	}
+
+	if err := runHAProxy(); err != nil {
+		log.Fatalf("Error running HAProxy: %s", err.Error())
+	}
+}
+
+func runHAProxy() error {
+	cmd := exec.Command("haproxy-systemd-wrapper", "-p", "/run/haproxy.pid",
```
Comment (ejj) [inline specs/haproxy/scripts/start.go]: What's this systemd wrapper?  Can we run haproxy directly without it?  Usually docker implies you don't need it.

At any rate, minimally deserves a comment.

File: specs/mean/Dockerfile [hunk 30]
```diff
@@ -0,0 +1,10 @@
...
+RUN mkdir -p /usr/src/app
+WORKDIR /usr/src/app
+
+RUN git clone https://github.com/tejasmanohar/node-todo.git .
```
Comment (ejj) [inline specs/mean/Dockerfile]: Btw, what's the status of the PR to upstream on this?

File: specs/mean/app.js [hunk 31]
```diff
@@ -0,0 +1,31 @@
+var image = "quilt/mean-service";
+
+function App(nWorker, port, env) {
```
Comment (ejj) [inline specs/mean/app.js]: I find this file a bit confusing.  What's App?  Is this a Node.js node?  I think the naming could be a tad more specific.

File: specs/mean/example.js [hunk 32]
```diff
@@ -0,0 +1,29 @@
+var HaProxy = require("github.com/NetSys/quilt/specs/haproxy/haproxy").Haproxy;
+var Mongo = require("github.com/NetSys/quilt/specs/mongo/mongo");
+var App = require("github.com/NetSys/quilt/specs/mean/app");
+
+// AWS
+var namespace = createDeployment({
+    namespace: "CHANGE_ME",
```
Comment (ejj) [inline specs/mean/example.js]: Ditch the CHANGE_ME

File: specs/mean/scripts/run.go [hunk 33]
```diff
@@ -0,0 +1,68 @@
...
+			log.Printf("Resolved %s to %+v", hostname, addrs)
+			return addrs
+		}
+
+		log.Printf("Unable to resolve %s: %s. Retrying...", hostname, err.Error())
+		time.Sleep(time.Second)
+	}
+}
+
+func waitForMongo(mongoURI string) {
```
Comment (ejj) [inline specs/mean/scripts/run.go]: Bleh.  Should probably file a trello card on this.  We'll need some way of expressing boot order dependencies at some point.

Comment (tejasmanohar) [inline specs/mean/scripts/run.go]: Ehhh, maybe I can add some node mongo driver settings to not timeout as quickly. I'm iffy about boot order since it doesn't make sense for updates which are the most common case

File: specs/mean/example.js (hunk 32 above)
Comment (tejasmanohar) [inline specs/mean/example.js]: made a JIRA ticket to do this everywhere

File: specs/mean/Dockerfile (hunk 30 above)
Comment (tejasmanohar) [inline specs/mean/Dockerfile]: sitting open, i'll email repo author

File: specs/haproxy/scripts/start.go (hunk 29 above)
Comment (tejasmanohar) [inline specs/haproxy/scripts/start.go]: copied it from old bash script but don't see how it's useful w docker since if haproxy exits unsuccessfully we just reboot the container. i'll swap it out

File: specs/haproxy/haproxy.js (hunk 28 above)
Comment (tejasmanohar) [inline specs/haproxy/haproxy.js]: good point, i'll follow jsdoc

**Comment by tejasmanohar** [comment]
Traveling over next few days. Will incorporate remaining feedback on Monday. If you don't mind reviewing anything left in meantime, that'd be great so hopefully we can merge by Tuesday :D
